feat: Port `Array.hpp` from CLP into a new `containers` library; Add initializer list constructor; Resolve most clang-tidy violations.

Bill-hbrhbr opened the pull request:
<!-- markdownlint-disable MD012 -->

<!--
Set the PR title to a meaningful commit message that:

* is in imperative form.
* follows the Conventional Commits specification (https://www.conventionalcommits.org).
  * See https://github.com/commitizen/conventional-commit-types/blob/master/index.json for possible
    types.

Example:

fix: Don't add implicit wildcards ('*') at the beginning and the end of a query (fixes #390).
-->

# Description

<!-- Describe what this request will change/fix and provide any details necessary for reviewers. -->

Create a new `containers` library in `ystdlib-cpp`, and port CLP's [`Array.hpp`](https://github.com/y-scope/clp/blob/3dc6927e09b0f7861da37f1023dcc9781bb4a063/components/core/src/clp/Array.hpp) as its first data structure class.

The purpose of the `Array` class can be found [here](https://github.com/y-scope/clp/pull/513).
As an enhancement, we add a new constructor that takes an initializer list, so that we can standardize the declaration/definition of compile-time fixed-sized arrays in CLP code, like [this](https://github.com/y-scope/clp/blob/60075f02a353e1009eaf2d0c9cb96b75d40b4269/components/core/src/clp/streaming_compression/lzma/Compressor.cpp#L167):
```cpp
    std::array<lzma_filter, 2> filters{{
            {.id = LZMA_FILTER_LZMA2, .options = &options},
            {.id = LZMA_VLI_UNKNOWN, .options = nullptr},
    }};
```
into
```cpp
    ystdlib::containers::Array<lzma_filter> filters{{
            {.id = LZMA_FILTER_LZMA2, .options = &options},
            {.id = LZMA_VLI_UNKNOWN, .options = nullptr},
    }};
```

The unit test [`test-Array.cpp`](https://github.com/y-scope/clp/blob/78c1c3fde7acb2b9f33ade152c13d04e290ade0c/components/core/tests/test-Array.cpp) is:
* refactored to remove code duplication
* aggressively modularized to eliminate clang-tidy's warnings about function complexity
* added with new tests to cover the new initializer list constructor

TODO: Consider supporting `Array` with non-default-initializable types in the future. This could be useful alongside the initializer list constructor but would require significant changes to the `Array` implementation. Since the benefits are minimal for now, an [issue](https://github.com/y-scope/ystdlib-cpp/issues/35) has been opened for tracking.

# Checklist

<!-- Ensure each item below is satisfied and indicate so by inserting an `x` within each `[ ]`. -->

* [X] The PR satisfies the [contribution guidelines][yscope-contrib-guidelines].
* [X] This is a breaking change and that has been indicated in the PR title, OR this isn't a
  breaking change.
* [X] Necessary docs have been updated, OR no docs need to be updated.

# Validation performed

<!-- Describe what tests and validation you performed on the change. -->

* [X] Added and passed new unit tests.


[yscope-contrib-guidelines]: https://docs.yscope.com/dev-guide/contrib-guides-overview.html


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Integrated a new container module into the library, now offering a flexible and robust Array utility with safe access and iteration support.
  - Enhanced the build configuration to include the new container component and extended test target options.

- **Tests**
  - Added comprehensive tests to verify the Array utility's functionality, covering various initialization methods, bounds checking, and basic operations, including edge cases.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Bill-hbrhbr commented:
`throw std::out_of_range("ystdlib::container::Array out-of-range access.");`
* Don't know if we want to apply an `ErrorCode` system for the `container` lib, or at least use `TraceableException` for this throw.
* Also don't know if we want to support initializing via `std::initializer_list`

Bill-hbrhbr commented:
@coderabbitai Explain 1-1. I've already ran clang-format. Point out where exactly the format is wrong

davidlion reviewed (changes requested):
1. improve the description
    * add to the description why we need the new constructor
    * mention that we're making a new library (container) and putting Array into it
2. change `container` to `containers`
3. come up with better names for `VarietyConstructors` and `NoDefault` classes in tests and rename `m_n`
    * I get that `n` is the size, but `m_n` is pretty confusing to read imo
    * potentially I just haven't read enough c++ recently and I'm more confused by the test file than I should be
4. since we're making changes to the libraries we port, is it worth trying to fix linting skips/ignores as well?
    * idk if the skips in Array are just fundamental to the design itself, but I feel like there is probably a way around it
5. try to re-write the title to match the style we've been using in PRs before someone else re-writes it
    * Array isn't a structure it is a class (or I didn't understand your meaning which is also a problem)
    * separate the point about adding a new library (container) and porting Array

Bill-hbrhbr commented:
@coderabbitai 177-181: Address the TODO comment about non-default-initializable types.
Create an issue to track

Bill-hbrhbr commented:
@coderabbitai
src/ystdlib/containers/test/test_Array.cpp (3)
1-14: Include headers could be organized by component group.
Yscope coding guidelines do not want to split standard libraries

Bill-hbrhbr commented:
@coderabbitai Yes that's what we are looking for

Bill-hbrhbr commented:
> 1. improve the description
>    
>    * add to the description why we need the new constructor
>    * mention that we're making a new library (container) and putting Array into it
> 2. change `container` to `containers`
> 3. come up with better names for `VarietyConstructors` and `NoDefault` classes in tests and rename `m_n`
>    
>    * I get that `n` is the size, but `m_n` is pretty confusing to read imo
>    * potentially I just haven't read enough c++ recently and I'm more confused by the test file than I should be
> 4. since we're making changes to the libraries we port, is it worth trying to fix linting skips/ignores as well?
>    
>    * idk if the skips in Array are just fundamental to the design itself, but I feel like there is probably a way around it
> 5. try to re-write the title to match the style we've been using in PRs before someone else re-writes it
>    
>    * Array isn't a structure it is a class (or I didn't understand your meaning which is also a problem)
>    * separate the point about adding a new library (container) and porting Array

1. PR description is improved with great details to cover all the changes.
   * Reason added for new constructor that takes intializer list
   * Explicitly separated the concept of adding `containers` and porting `Array.hpp`
2. Changed
3. Renamed as follows:
   * `VarietyConstructors` to `PolymorphicConstructors`
   * `NoDefault` to `ExplicitConstructor` (also added why there is this class in a TODO)
   * `m_n` to `m_value`
4. I've tried to remove as many `NOLINT` comments as possible, but the remaining ones are fundamentally unable to be adapted.
5. Rewritten.

Bill-hbrhbr commented:
@coderabbitai
src/ystdlib/containers/test/test_Array.cpp (4)
17-19: Use namespaced constants for better scoping.
For yscope coding guidelines, file-scope constants/helper functions should always be in anon namespace.

150-152: Verify zero initialization for all types.
Comment is unnecessary until proven wrong that a value init is not 0

158-205: Todo comment needs to be addressed or tracked.

Already tracked https://github.com/y-scope/ystdlib-cpp/issues/35

Bill-hbrhbr commented:
@coderabbitai 
201-204: Consider documenting the limitation in Array.hpp.
There's already doc and explicit constraint
```cpp
namespace ystdlib::containers {
/**
 * Class for a runtime fix-sized array.
 * @tparam T The type of elements in the array. The type must be default initializable so that this
 * class doesn't need to implement special memory allocation procedure for std::unique_ptr<T[]>.
 * Note that when using constructor that takes an initializer list, T must be copy constructable.
 */
template <typename T>
requires(std::is_fundamental_v<T> || std::default_initializable<T>)
class Array {
```

Bill-hbrhbr commented:
@coderabbitai
src/ystdlib/containers/test/test_Array.cpp (3)
85-92: Verify that both mutable and const references work correctly.
it's too thorough we skip it this time

104-113: Good test for move semantics.
it's in a valid but unspecified state. Should not assume the sizes

151-199: Thorough testing of list initialization capabilities.
Stop asking if you want to help me implement TODOs. They are left as TODOs because we don't want to implement them in this PR.

davidlion commented on src/ystdlib/containers/test/test_Array.cpp:
```diff
@@ -0,0 +1,200 @@
+
+    // Test polymorphic list initialization with a non default-constructible class
+    REQUIRE_FALSE(std::default_initializable<ExplicitConstructor>);
+
+    // TODO: Test `Array` list initialization with this class. Currently, `Array` requires the
+    // template type to be a fundamental type or a default-initializable class type.
```
```suggestion
```
